### ISSUE TITLE
1.8.1 breaks simple XML parsing

### ISSUE THREAD

**Issue by GaetanLepage** [crystal-lang/crystal#13444]
## Bug Report

Hello,
The following code works fine witch crystal 1.8.0 but breaks with 1.8.1:
```crystal
str = <<-XML
<foo>ThisIsFoo</foo>
<bug:bar>ThisIsBar</bug:bar>
XML

require "xml"

rss = XML.parse_html(str)

foo = rss.xpath_node("//foo").not_nil!
pp foo.content

bar = rss.xpath_node("//bar").not_nil!
pp bar.content
```

Crystal 1.8.0:
```
"ThisIsFoo"
"ThisIsBar"
```

Crystal 1.8.1:
```
"ThisIsFoo"
Unhandled exception: Nil assertion failed (NilAssertionError)
  from /usr/lib/crystal/nil.cr:113:7 in 'not_nil!'
  from /usr/lib/crystal/nil.cr:109:3 in 'not_nil!'
  from eval:13:7 in '__crystal_main'
  from /usr/lib/crystal/crystal/main.cr:115:5 in 'main_user_code'
  from /usr/lib/crystal/crystal/main.cr:101:7 in 'main'
  from /usr/lib/crystal/crystal/main.cr:127:3 in 'main'
  from /usr/lib/libc.so.6 in '??'
  from /usr/lib/libc.so.6 in '__libc_start_main'
  from /home/crystal/.cache/crystal/crystal-run-eval.tmp in '_start'
  from ???
```

**Comment by Blacksmoke16** [comment]
What version of `libxml2` do you have installed? I'm thinking this is more so due to changes in the underlying linked/installed lib versus Crystal itself. Given it reproduces for me `1.7.3` as well.

**Comment by SamantazFox** [comment]
I tried with different containers, it breaks with containers tagged `1.8.x-alpine` but not with `1.8.x`

As discussed over Matrix, might be caused by the underlying libxml version:
```
$ echo 'FROM docker.io/crystallang/crystal:1.8.0-alpine \nCMD ["apk", "list", "-i"]' | docker build -f - -t crystal-test && docker run crystal-test | grep xml
[snip]
libxml2-2.10.4-r0 x86_64 {libxml2} (MIT) [installed]
libxml2-static-2.10.4-r0 x86_64 {libxml2} (MIT) [installed]
libxml2-utils-2.10.4-r0 x86_64 {libxml2} (MIT) [installed]
libxml2-dev-2.10.4-r0 x86_64 {libxml2} (MIT) [installed]
```
```
$ echo 'FROM docker.io/crystallang/crystal:1.8.0 \nCMD ["apt", "list", "--installed"]' | docker build -f - -t crystal-test && docker run crystal-test | grep xml
[snip]
libxml2-dev/now 2.9.13+dfsg-1ubuntu0.2 amd64 [installed,local]
libxml2/now 2.9.13+dfsg-1ubuntu0.2 amd64 [installed,local]
```

**Comment by SamantazFox** [comment]
https://gitlab.gnome.org/GNOME/libxml2/-/compare/v2.10.4...v2.9.14?from_project_id=1665&straight=false

**Comment by Blacksmoke16** [comment]
On second thought, we could possibly consider re-building/distributing new images/binaries with a patched version of libXML if we think that would be worthwhile. Granted we'll probably be limited on the available versions within the base images, and/or depending on how Crystal was installed might be up to the user to upgrade the version on their own machine.

**Comment by straight-shoota** [comment]
I would not override the libXML decision. If they think this breaking change is justified, I'm fine with that. The only reason for us to do something about it would be if it breaks some legitimate use case withing the Crystal bindings.
This is not the case here because `parse_html` was incorrectly used for a document that is *not* HTML. Using `XML.parse` instead is the correct fix (https://github.com/iv-org/invidious/pull/3793).
Perhaps it would be nice to have more convenient way to ignore namespaces when querying the document if you don't really care about the specifics.

**Comment by SamantazFox** [comment]
> TL;DR: Your link only shows the commits which were added to `v2.9.14` since `v2.9.13`.
> 
> This link shows the actual changes between the two versions: https://gitlab.gnome.org/GNOME/libxml2/-/compare/v2.9.13...v2.10.4?from_project_id=1665&straight=false

Ah, thanks a bunch fro the correction!

@GaetanLepage mentioned the following issue on our Matrix chat, but we forgot to post it here:
https://gitlab.gnome.org/GNOME/libxml2/-/issues/508

> On second thought, we could possibly consider re-building/distributing new images/binaries with a patched version of libXML if we think that would be worthwhile.

After thinking more about it (a whole night of rest makes wonders!), I second @straight-shoota 's answer.
Our usage was not the best, and we should have been using `XML.parse` from the start.

Also, the recent CVEs hitting libxml2 should be a reminder of the criticality of this library.
Maintaining a separate branch is (in my opinion) nothing but a dangerous game!

**Comment by straight-shoota** [comment]
Okay so it seems the change in libXML was not deliberate and should be fixed in a newer release?

That still doesn't mean we should do anything about it. It's possible to get the desired behaviour by using an older or a patched version of libXML2.

**Comment by HertzDevil** [comment]
As I have also noted in https://github.com/crystal-lang/crystal/issues/13273#issuecomment-1508497948, libxml2 does not define ways to retrieve its version during execution time. You could detect it at compile-time via other means (`xml2-config` or `pkg-config`?) but it will never be as robust as the runtime checks, as we recently did with PCRE2.
